File: python/design-patterns/template/template/beverage_maker.py
from abc import ABC, abstractmethod

class BeverageMaker(ABC):
    # The Template Method
    def prepare_recipe(self):
        self._boil_water()
        self._brew()
        self._pour_in_cup()
        self._add_condiments()

    def _boil_water(self): print("Boiling water to 95C")
    def _pour_in_cup(self): print("Pouring in cup")

    @abstractmethod
    def _brew(self): pass
    @abstractmethod
    def _add_condiments(self): pass

File: typescript/design-patterns/template/template/BeverageMaker.ts
// Abstract Base: Defines the non-overridable template method
abstract class BeverageMaker {
  // 🎯 The Template Method: locks in the step sequence
  public final prepareRecipe(): void {
    this.boilWater();
    this.brew();
    this.pourInCup();
    this.addCondiments();
  }

  private boilWater() {
    console.log("Boiling water to 95°C...");
  }

  private pourInCup() {
    console.log("Pouring beverage into standard cup...");
  }

  // Abstract hooks to be customized by subclasses
  protected abstract brew(): void;
  protected abstract addCondiments(): void;
}